Domains: Update domain mapping thank you page

leonardost opened the pull request:
#### Changes proposed in this Pull Request

In order to improve the user experience when buying domains and mappings, domain mapping instructions were updated in #50668. This PR updates the `Thank You` page, which is shown after a user buys a domain mapping, to reflect the same instructions that were changed in #50668.

Depending on the type of site the mapping was bought for - root domain or sub domain, for a simple or AT site - there should be one recommended setup (usually setting name servers) and one or two alternative, more advanced, options, usually setting A or CNAME records.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->
Both simple and atomic sites should be tested, as well as root domain mappings and subdomain mappings.

1. In Calypso, go to Upgrades > Domains > Add a domain to this site
2. Select "Use a domain I own"
3. "Map your domain"
4. Insert a root domain or subdomain
5. Buy that domain mapping for your site
6. The "Thank you" page with further instructions to setup domain mapping will be shown.

Root domain mappings instructions differ if the site is simple or atomic.
- For simple sites: The recommended setup should be to set NS records pointing to our name servers, and the advanced setup should show the default IP addresses (*.24 and *.25) to be set as A records
- For AT sites: The recommended setup should be to set NS records pointing to our name servers, and the advanced setup should show the AT site's IP addresses to be set as A records

Subdomain mappings should show the following instructions: The recommended setup should be to set CNAME or NS records pointing to our name servers, and no advanced setup should be shown

leonardost committed:
Update PurchaseDetail to show button only if a href is given

leonardost committed:
Update domain mapping thank you page instructions

Update instrucitons to be the same as those defined in #50668.

Only root domain mapping instructions were updated in this commit, and
A records are still not being shown.

leonardost committed:
Show A records in domain mapping thank you page instructions

klimeryk commented on client/my-sites/checkout/checkout-thank-you/domain-mapping-details.jsx:
```diff
@@ -10,128 +10,128 @@ import { localize } from 'i18n-calypso';
-	) : (
-		<span />
-	);
+	const renderLinkTo = ( url ) => {
```
Let's use `ExternalLink` here.

klimeryk commented on client/my-sites/checkout/checkout-thank-you/domain-mapping-details.jsx:
```diff
@@ -10,128 +10,128 @@ import { localize } from 'i18n-calypso';
+	const primaryMessage = translate(
+		'Please log into your account at your domain registrar and {{strong}}update the name servers{{/strong}} of your domain to use the following values, as detailed in {{link}}these instructions{{/link}}:',
+		{
+			comment: 'Notice for mapped domain notice with NS records pointing to somewhere else',
```
This might not be needed.

klimeryk commented on client/my-sites/checkout/checkout-thank-you/domain-mapping-details.jsx:
```diff
@@ -10,128 +10,128 @@ import { localize } from 'i18n-calypso';
-					<li>{ domain }. IN NS ns1.wordpress.com.</li>
-					<li>{ domain }. IN NS ns2.wordpress.com.</li>
-					<li>{ domain }. IN NS ns3.wordpress.com.</li>
+	const isDataLoaded = () => {
```
Let's maybe refactor the whole component to be a normal one - or at least, pull these methods out of this component.

klimeryk commented on client/my-sites/checkout/checkout-thank-you/domain-mapping-details.jsx:
```diff
@@ -10,128 +10,128 @@ import { localize } from 'i18n-calypso';
+	};
+
+	const renderARecordsList = () => {
+		if ( isDataLoaded() ) {
```
Maybe let's inverse the logic here - and return the placeholder component early. As that's the least interesting case (and avoids big nested lines).

klimeryk commented on client/my-sites/checkout/checkout-thank-you/style.scss:
```diff
@@ -53,6 +53,27 @@
+		margin: 0;
+	}
+	.purchase-detail__image {
+		@include breakpoint-deprecated( '>660px' ) {
```
Let's try to use the new Gutenberg mixins in new code: https://github.com/Automattic/wp-calypso/blob/HEAD/docs/coding-guidelines/css.md#media-queries
And https://github.com/WordPress/gutenberg/blob/0f1f5e75408705f0ec014f5d2ea3d9fcc8a97817/packages/base-styles/_breakpoints.scss

klimeryk reviewed (commented):
Leaving comments from our call together so that they are not lost.
Sorry for the brevity - the code and approach is overall looking really good. Good job figuring out all the moving pieces from other places in our code!

leonardost committed:
Change `a` tag for `ExternalLink` component

leonardost committed:
Remove redundant comment from translation string

leonardost committed:
Invert conditions in renderARecordsList method to improve readability

leonardost committed:
Remove unnecesary CSS rule

leonardost committed:
Replace deprecated mixin for break-mobile

leonardost committed:
Extract some functions from DomainMappingDetails component

leonardost committed:
Add subdomain mapping instrucitons to mappings thank you page

leonardost committed:
Extract mapping instructions into DomainMappingInstructions component

klimeryk commented on client/my-sites/domains/domain-management/domain-mapping-instructions/index.jsx:
```diff
@@ -0,0 +1,143 @@
+		domain: PropTypes.object,
+		isAtomic: PropTypes.bool,
+		isLoaded: PropTypes.bool,
+		translate: PropTypes.func.isRequired,
```
By convention, `translate` is not included in the prop types, as it comes from the `localize` HOC wrapper.

klimeryk commented on client/my-sites/domains/domain-management/domain-mapping-instructions/index.jsx:
```diff
@@ -0,0 +1,143 @@
+		isLoaded: false,
+	};
+
+	render() {
```
By convention, `render` is usually the last method in the component. The way I've always approached it was go from most internal to most "public" ones. So, lifecycle methods like `willReceiveProps`, etc. first and then smaller `render*` methods. And then finally the main `render` method.

klimeryk commented on client/my-sites/domains/domain-management/domain-mapping-instructions/index.jsx:
```diff
@@ -0,0 +1,143 @@
+	};
+
+	render() {
+		const { domain } = this.props;
```
To avoid inconsistent messaging as the `domain` object is loaded, let's maybe simply show a placeholder here and return early.

klimeryk commented on client/my-sites/domains/domain-management/domain-mapping-instructions/index.jsx:
```diff
@@ -0,0 +1,143 @@
+	static propTypes = {
+		domain: PropTypes.object,
+		isAtomic: PropTypes.bool,
+		isLoaded: PropTypes.bool,
```
Let's maybe put a more descriptive variable name here - it's unclear what exactly "is loaded" otherwise.

klimeryk commented on client/my-sites/domains/domain-management/domain-mapping-instructions/index.jsx:
```diff
@@ -0,0 +1,143 @@
+
+class DomainMappingInstructions extends React.Component {
+	static propTypes = {
+		domain: PropTypes.object,
```
Instead of passing in the whole domain object, maybe it'd be better to pass in only the props we need:

 - domain name
 - A records list
 - `areDomainDetailsLoaded` type of prop that will probably look like `! isRequestingSiteDomains && domain` - meaning, do we have the domain object and we're _not_ waiting for a new version of it.

klimeryk commented on client/my-sites/domains/domain-management/domain-mapping-instructions/index.jsx:
```diff
@@ -0,0 +1,143 @@
+/**
```
As discusses on call, maybe let's put it in `my-sites/domains/components/mapping-instructions`, since this will be shared between domain management and checkout :bow:

klimeryk commented on client/my-sites/checkout/checkout-thank-you/domain-mapping-details.jsx:
```diff
@@ -10,128 +10,162 @@ import { localize } from 'i18n-calypso';
 	const selectedSite = getSelectedSite( state );
 	return {
+		domains: getDomainsBySiteId( state, selectedSite.ID ),
 		isBusinessPlan: isBusiness( selectedSite.plan ),
```
Probably need to add something like `isRequestingSiteDomains: isRequestingSiteDomains( state, selectedSite.ID )` here from `import { isRequestingSiteDomains } from 'calypso/state/sites/domains/selectors';`.

klimeryk reviewed (commented):
Overall, I really like the separate component that we will be then be able to share in different places. Left a few comments from our call.

Be sure to also test with `?flag=force-sympathy` in development mode to simulate loading Calypso with an empty Redux tree.

leonardost committed:
Move CSS rule from DomainMappingDetails to DomainMappingInstructions

leonardost committed:
Remove DomainMappingInstructions code from DomainMappingDetails

leonardost committed:
Remove unneeded `translate` from PropTypes

leonardost committed:
Invert order of render* methods to adhere to convention

leonardost committed:
Change location of DomainMappingInstructions component

Moved from `my-sites/domains/domain-management/` to
`my-sites/domains/components/` as the component will be shared between
doman management and checkout.

leonardost committed:
Alter component folder name from `domain-mapping-instructions` to `mapping-instructions`

leonardost committed:
Rename attribute `isLoaded` to `areDomainDetailsLoaded`

leonardost committed:
Pass `aRecords` and `domainName` properties to `DomainMappingInstrucitons`

leonardost committed:
Update `areDomainDetailsLoaded` logic to check if domain details were already fetched

leonardost committed:
Show placeholder for DomainMappingInstructions component while domain details are not loaded

leonardost committed:
Reduce padding around mapping instructions in thank you page for small screens

leonardost committed:
Add padding around mapping instrucitons placeholder

klimeryk commented on client/my-sites/checkout/checkout-thank-you/domain-mapping-details.jsx:
```diff
@@ -10,128 +10,52 @@ import { localize } from 'i18n-calypso';
 const mapStateToProps = ( state, { domain } ) => {
 	const selectedSite = getSelectedSite( state );
 	return {
+		domains: getDomainsBySiteId( state, selectedSite.ID ),
```
```suggestion
	const selectedSite = getSelectedSite( state );
	const domains = getDomainsBySiteId( state, selectedSite.ID );
	return {
		purchasedDomain: getSelectedDomain( { domains, selectedDomainName: domain } );
```

I'd move fetching the selected domain name to here, instead of doing it inside the component.

klimeryk commented on client/my-sites/checkout/checkout-thank-you/domain-mapping-details.jsx:
```diff
@@ -10,128 +10,52 @@ import { localize } from 'i18n-calypso';
+import DomainMappingInstructions from 'calypso/my-sites/domains/components/mapping-instructions';
 
 const DomainMappingDetails = ( {
 	domain,
```
Maybe to avoid confusion on whether `domain` is the domain object or string, you could rename it here like this:

```suggestion
	domain: domainName,
```

klimeryk commented on client/components/purchase-detail/index.jsx:
```diff
@@ -78,7 +78,7 @@ export default class PurchaseDetail extends PureComponent {
 		return (
 			<div className="purchase-detail__body">
-				{ this.renderPurchaseButton() }
+				{ this.props.href && this.renderPurchaseButton() }
```
This change might not be needed or correct in fact - I can imagine a use case where `onClick` is the only handler passed (like navigating to a different section in Calypso) and this change would break that behaviour.

Maybe we're misusing this component at all - instead of passing in `description`, it should not be a `body`?

klimeryk reviewed (commented):
Awesome! This looks almost perfect - I've just left a few nitpicks, but we're almost there ❤️

leonardost committed:
Move `purchasedDomain` attribute inside `mapStateToProps`

leonardost committed:
Rename `domain` property to `selectedDomainName`

leonardost committed:
Revert unnecessary check in PurchaseDetail to hide button

The removed check was redundant.

leonardost committed:
Use `body` attribute of PurchaseDetail instead of `description`

leonardost committed:
Fix identation of DomainMappingInstructions' style.scss file

leonardost committed:
Replace mapping instructions in MappedDomainType by the DomainMappingInstructions component

leonardost committed:
Remove conflicting CSS rule from domain-management/edit/style.scss

leonardost commented on client/my-sites/domains/domain-management/edit/style.scss:
```diff
@@ -165,10 +165,6 @@
-
-			> div:last-of-type .foldable-faq__answer {
-				border-bottom: 0;
-			}
```
This rule was conflicting with another one in `mapping-instructions/style.scss`

klimeryk reviewed (approved):
Tested again and all looks great! 💖

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/5639196

Hi @leonardost, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include this string: `Please set the following IP addresses as A records for %(subdomain)s using {{link}}these instructions{{/link}}:`

Thank you in advance!

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/5639196).
